main.go

```diff
@@ -44,8 +56,163 @@ func usage(out io.Writer) {
 	fmt.Fprintf(out, "  -k, --no-verify      Do not verify TLS/SSL certificates.\n")
 	fmt.Fprintf(out, "  -r, --redirect       Rewrite and return 3xx redirects.\n")
 	fmt.Fprintf(out, "      --keep-referer   Pass Referer: headers through, even with -r.\n")
+	fmt.Fprintf(out, "      --tls            Present TLS (with a custom CA) to clients connecting\n")
+	fmt.Fprintf(out, "                       to us.  The CA will be dumped to standard error.\n")
+}
+
+type Cert struct {
+	RawCertificate *x509.Certificate
+	RawKey         *rsa.PrivateKey
+
+	Certificate string
+	Key         string
+}
+
+func (ca *Cert) Sign(cert *Cert) error {
+	raw, err := x509.CreateCertificate(rand.Reader, cert.RawCertificate, ca.RawCertificate, cert.RawKey.Public(), ca.RawKey)
+	if err != nil {
+		return err
+	}
+
+	cert.Certificate = string(pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: raw,
+	}))
+
+	return nil
+}
+
+func certificate(name string, serial int, ttl time.Duration) (*Cert, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	cert := &x509.Certificate{
+		SignatureAlgorithm:    x509.SHA512WithRSA, /* FIXME: hard-coded */
+		PublicKeyAlgorithm:    x509.RSA,
+		Subject:               pkix.Name{CommonName: name},
+		DNSNames:              []string{name},
+		IPAddresses:           []net.IP{net.ParseIP("127.0.0.1")},
+		BasicConstraintsValid: true,
+		IsCA:         true,
+		MaxPathLen:   1,
+		SerialNumber: big.NewInt(1),
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().Add(ttl),
+	}
+
+	pKey := string(pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	}))
+
+	return &Cert{
+		RawCertificate: cert,
+		RawKey:         key,
+
+		//	Certificate: pCert,
+		Key: pKey,
+	}, nil
 }
 
+func loadOrGenerateCA() (*Cert, error) {
+	if _, err := os.Stat(GotchaCAPath); !os.IsNotExist(err) {
```

Comment:
A better memoization pattern is to check that the cached copy does _not_ exist, and if so, generate the cache.  Then the rest of the function can read from the cache.

It seems a tad wasteful for the initial case, but it does mean that, no matter what, we're always loading the CA cert/key via the same code paths, and you don't end up with weird bugs where it works the first time but fails on subsequent calls because the gen-and-cache logic is writing something it can't parse later.